### ISSUE TITLE
SocketException: No such host is known

### ISSUE THREAD

**Issue by borrelnoten** [alanmcgovern/monotorrent#404]
nuget package 2.0 instable rev 0122 version. Not sure which torrent it was giving this exception.

Exception thrown: 'MonoTorrent.Client.Tracker.TrackerException' in MonoTorrent.dll
Scrape could not be completed

**This exception was originally thrown at this call stack:**
    System.Net.Dns.GetAddrInfo(string)
    System.Net.Dns.InternalGetHostByName(string, bool)
    System.Net.Dns.GetHostAddresses(string)
    System.Net.Sockets.UdpClient.Connect(string, int)
    System.Net.Sockets.UdpClient.UdpClient(string, int)
    MonoTorrent.Client.Tracker.UdpTracker.SendAndReceiveAsync(MonoTorrent.Client.Messages.UdpTracker.UdpTrackerMessage)
    System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
    System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
    MonoTorrent.Client.Tracker.UdpTracker.ConnectAsync()
    System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)

**Exception:**

```
{"Scrape could not be completed"}
    Data: {System.Collections.ListDictionaryInternal}
    HResult: -2146233088
    HelpLink: null
    IPForWatsonBuckets: 0x71f8cc51
    InnerException: {"No such host is known"}
    IsTransient: false
    Message: "Scrape could not be completed"
    RemoteStackTrace: null
    Source: "MonoTorrent"
    StackTrace: "   at MonoTorrent.Client.Tracker.UdpTracker.<DoScrapeAsync>d__14.MoveNext()"
    TargetSite: {Void MoveNext()}
    WatsonBuckets: null
    _HResult: -2146233088
    _className: null
    _data: {System.Collections.ListDictionaryInternal}
    _dynamicMethods: null
    _exceptionMethod: null
    _exceptionMethodString: null
    _helpURL: null
    _innerException: {"No such host is known"}
    _ipForWatsonBuckets: 0x71f8cc51
    _message: "Scrape could not be completed"
    _remoteStackIndex: 0
    _remoteStackTraceString: null
    _safeSerializationManager: {System.Runtime.Serialization.SafeSerializationManager}
    _source: "MonoTorrent"
    _stackTrace: {sbyte[24]}
    _stackTraceString: null
    _watsonBuckets: null
    _xcode: -532462766
    _xptrs: 0x00000000
```

**Inner exception:**

```
{"No such host is known"}
    Data: {System.Collections.ListDictionaryInternal}
    ErrorCode: 11001
    HResult: -2147467259
    HelpLink: null
    IPForWatsonBuckets: 0x71f8cc51
    InnerException: null
    IsTransient: false
    Message: "No such host is known"
    NativeErrorCode: 11001
    RemoteStackTrace: null
    SocketErrorCode: HostNotFound
    Source: "System"
    StackTrace: "   at System.Net.Dns.GetAddrInfo(String name)\r\n   at System.Net.Dns.InternalGetHostByName(String hostName, Boolean includeIPv6)\r\n   at System.Net.Dns.GetHostAddresses(String hostNameOrAddress)\r\n   at System.Net.Sockets.UdpClient.Connect(String hostname, Int32 port)\r\n   at System.Net.Sockets.UdpClient..ctor(String hostname, Int32 port)\r\n   at MonoTorrent.Client.Tracker.UdpTracker.<SendAndReceiveAsync>d__16.MoveNext()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at MonoTorrent.Client.Tracker.UdpTracker.<ConnectAsync>d__15.MoveNext()\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at MonoTorrent.Client.Tracker.UdpTracker.<DoScrapeAsync>d__14.MoveNext()"
    TargetSite: {System.Net.IPHostEntry GetAddrInfo(System.String)}
    WatsonBuckets: null
    _HResult: -2147467259
    _className: null
    _data: {System.Collections.ListDictionaryInternal}
    _dynamicMethods: null
    _exceptionMethod: null
    _exceptionMethodString: null
    _helpURL: null
    _innerException: null
    _ipForWatsonBuckets: 0x71f8cc51
    _message: "No such host is known"
    _remoteStackIndex: 0
    _remoteStackTraceString: null
    _safeSerializationManager: {System.Runtime.Serialization.SafeSerializationManager}
    _source: "System"
    _stackTrace: {sbyte[384]}
    _stackTraceString: null
    _watsonBuckets: null
    _xcode: -532462766
    _xptrs: 0x00000000
    m_EndPoint: null
    nativeErrorCode: 11001
```

**Comment by borrelnoten** [comment]
seems it is directly related to the trackers I addd to a magnet URL.
Could be some of these trackers are bad and result in the above issues.

**Comment by SICGames** [comment]
Are you getting a MagnetUrl or from a torrent file? If a magnet file, I always use 

`System.Web.HttpUtility.UrlDecode(url);`

The JSON data that I get back throws in EncodedUrl stuff. Also, for test purposes I added trackers from a file and made sure to use the HttpUtility. 

Not sure if this will help. But if you're obtaining a magnet url. This breaks everything down. Maybe not very optimized. 

```
    public static class TorrentMagnetUtils
    {
        public static MonoTorrent.MagnetLink ConvertMagnetUrl(string magnetUrl)
        {
            
            magnetUrl = System.Web.HttpUtility.UrlDecode(magnetUrl);
            var hash = magnetUrl.Substring(magnetUrl.IndexOf("btih:") + 5);
            hash = hash.Substring(0, hash.IndexOf("&dn"));
            var downloadName = magnetUrl.Substring(magnetUrl.IndexOf("&dn=") + 4);
            downloadName = downloadName.Substring(0, downloadName.IndexOf("&tr"));
            List<String> announcements = new List<string>();

            MatchCollection mc = Regex.Matches(magnetUrl, @"(\?|\&)([^=]+)\=([^&]+)", RegexOptions.Singleline);

            foreach (Match match in mc)
            {

                string value = match.Value;
                if (value.Contains("&tr="))
                {
                    string announcement = value.Substring(match.Value.IndexOf("&tr=") + 4);
                    announcements.Add(announcement);
                }
            }

            return new MonoTorrent.MagnetLink(MonoTorrent.InfoHash.FromHex(hash), downloadName, announcements);
        }
    }

```

What's the main magnet url by the way? I can probably test out those trackers. The trackers i've added to the file is quite long. But I do recognize some of your trackers that you've used. And, no issues so far.

**Comment by alanmcgovern** [comment]
This looks like the normal behaviour when a given Tracker cannot be contacted. 

If you discovered this exception by virtue of running the code with the visual studio debugger attached then there's no further diagnosing needed unless you have reason to believe something more serious is going on. 

I'll close this for now as this should be fine.

> System.Web.HttpUtility.UrlDecode(url)

I'd recommend using this only if the string has been url encoded in the first place. It would be wrong to do this on a value which isn't url encoded.

Additionally - it's odd to see regex being used to parse such a simple url format. When aren't you simply parsing the string using the approriate static method on the MagnetLink class?

**Comment by SICGames** [comment]
> This looks like the normal behaviour when a given Tracker cannot be contacted.
> 
> If you discovered this exception by virtue of running the code with the visual studio debugger attached then there's no further diagnosing needed unless you have reason to believe something more serious is going on.
> 
> I'll close this for now as this should be fine.
> 
> > System.Web.HttpUtility.UrlDecode(url)
> 
> I'd recommend using this only if the string has been url encoded in the first place. It would be wrong to do this on a value which isn't url encoded.
> 
> Additionally - it's odd to see regex being used to parse such a simple url format. When aren't you simply parsing the string using the approriate static method on the MagnetLink class?

the magnetlink url obtained is from a JSON response from a website. And it has encoded URL. Before, I tried to use the MagnetLink after manually parsing the hashInfo. However, it didn't behave like I wanted. Maybe I didn't know how to use it. 

magnetlink obtained from JSON response:

> magnet:?xt=urn:btih:0477C03A7C27555BEE24F123D774662ADE08D23F&dn=Tangled (2010) 1080p&tr=http://retracker.uln-ix.ru/announce&tr=http://tracker.best-torrents.net:6969/announce&tr=http://91.218.230.81:6969/announce&tr=http://retracker.telecom.kz/announce&tr=http://retracker.adminko.org/announce&tr=http://tracker.tfile.me/announce&tr=http://mgtracker.org:2710/announce&tr=http://bt.careland.com.cn:6969/announce

I optimized the MagnetLinkConverter function to parse the url within 0.000300 ms.  Oh, that isn't the raw response. It's pretty versioned. 

If you have a MagnetLink parser, i'll try it out. I've been constantly using the newest MonoTorrent and it's purring very nicely.

**Comment by alanmcgovern** [comment]
The spec compliant (url encoded) form for this magnet link would be: 
```
magnet:?xt=urn:btih:0477C03A7C27555BEE24F123D774662ADE08D23F&dn=Tangled+(2010)+1080p&tr=http%3a%2f%2fretracker.uln-ix.ru%2fannounce&tr=http%3a%2f%2ftracker.best-torrents.net%3a6969%2fannounce&tr=http%3a%2f%2f91.218.230.81%3a6969%2fannounce&tr=http%3a%2f%2fretracker.telecom.kz%2fannounce&tr=http%3a%2f%2fretracker.adminko.org%2fannounce&tr=http%3a%2f%2ftracker.tfile.me%2fannounce&tr=http%3a%2f%2fmgtracker.org%3a2710%2fannounce&tr=http%3a%2f%2fbt.careland.com.cn%3a6969%2fannounce
```

And you can parse this by passing it to `MagnetLink.Parse (string)`.

One real world example where UrlDecoding will break the magnet link would be display names like `Pride & Prejudice`. The correct magnetlink for this is: `magnet:?xt=urn:btih:0477C03A7C27555BEE24F123D774662ADE08D23F&dn=Pride+%26+Prejudice` and `MagnetLink.Parse` will decode this back into `Pride & Prejudice`.

If you choose to UrlDecode this before passing it to `MagnetLink.Parse` then you would be attempting to parse the string `"magnet:?xt=urn:btih:0477C03A7C27555BEE24F123D774662ADE08D23F&dn=Pride & Prejudice"`, and when you split the query string on `&` as required by normal url parsing rules you'll end up corrupting the display name.

In short - either use `MagnetLink.Parse` and everything will Just Work, or manually UrlDecode the string and parse it by hand. should *not* manually UrlDecode* the string *and also* use `MagnetLink.Parse` :)
